renku/core/management/workflow/providers/toil.py

```diff
@@ -77,19 +77,23 @@ def run(self, storage):
         for p in self._parents_promise:
             parent_inputs.update(p)
 
-        for i in self.workflow.inputs:
-            file_metadata = (
-                parent_inputs[i.actual_value] if i.actual_value in parent_inputs else self._input_files[i.actual_value]
-            )
-            input_path = Path(i.actual_value)
+        def _read_input(input: Path, file_metadata):
+            input_path = Path(input)
+
             if isinstance(file_metadata, dict):
-                input_path.mkdir(exist_ok=True)
+                input_path.mkdir(parents=True, exist_ok=True)
                 for path, file_id in file_metadata.items():
-                    storage.readGlobalFile(file_id, userPath=path)
-            else:
+                    _read_input(path, file_id)
+            elif not input_path.exists():
```

Comment:
This won't `readGlobalFile` if the `input_path` exists; is this the correct behavior?

Reply:
yes that's intentional. `readGlobalFile` has no option for overwrite, i.e. it throws an exception if `userPath` already exists; this avoid pulling the given file from the global storage if it's already available.